Add support for UTF8 filenames in wxZipOutputStream

TcT2k opened the pull request:
Zip filenames containing non ASCII characters will be marked with
bit 11 in the general purpose flags and will use UTF-8 encoding.

By only setting the flag when non ASCII characters are used the
created archives should be binary identical to previous versions.

The old behavior can be achieved by explicitly using a wxMBConv
object with the constructor. This should also ensure that
existing code using a custom wxMBConv should work as before.

_This probably needs a few unit tests._

TcT2k committed:
Add support for UTF8 filenames in wxZipOutputStream

Zip filenames containing non ASCII characters will be marked with
bit 11 in the general purpose flags and will use UTF-8 encoding.

By only setting the flag when non ASCII characters are used the
created archives should be binary identical to previous versions.

The old behavior can be achieved by explicitly using wxConvLocal
with the constructor. This should also ensure that
existing code using a custom wxMBConv should work as before.

vadz commented on interface/wx/zipstrm.h:
```diff
@@ -476,14 +476,34 @@ class wxZipOutputStream : public wxArchiveOutputStream
+        UTF-8 Encoding will be used for filenames that are using non ASCII
+        characters.
+
+        @since 3.1.1
```
I think this risks being misleading in the generated documentation, i.e. it would create an impression that it's the ctor itself which is new, which is not really the case as the code using it would still compile with older versions and would just work differently. It's probably better to just write in plain text that "Since versions 3.1.1, filenames in the generated archive will be encoded using UTF-8" and ideally also explain how can it be changed if really needed (although I have no idea why would anyone need to change it to be honest).

I also think it might be worth pretending that there is still only a single ctor (for each pointer/reference variant) instead of 2 overloaded ones, even if we don't actually have a default value for `conv` parameter, it would IMHO be simpler to understand if we still documented it like this, with `wxConvUTF8` being the default.

vadz commented on docs/changes.txt:
```diff
@@ -74,6 +74,9 @@ INCOMPATIBLE CHANGES SINCE 3.1.0:
 - The enum value wxTASKBAR_JUMP_LIST_DESTIONATION, which was added in 3.1.0,
   contains a typo and has been renamed to wxTASKBAR_JUMP_LIST_DESTINATION.
+  
+- wxZipOutputStream will now automatically convert filenames to UTF-8, if no
```
I'd say "... will now use UTF-8 to encode filenames and comments in generated archive instead of the current locale encoding by default".

vadz commented on include/wx/zipstrm.h:
```diff
@@ -225,17 +225,19 @@ class WXDLLIMPEXP_BASE wxZipEntry : public wxArchiveEntry
 
     size_t ReadLocal(wxInputStream& stream, wxMBConv& conv);
-    size_t WriteLocal(wxOutputStream& stream, wxMBConv& conv) const;
+    size_t WriteLocal(wxOutputStream& stream, wxMBConv& conv, bool checkForUTF8) const;
```
As always, I'd prefer to use something like `int flags = Flags_UseUTF8IfNecessary` instead of `bool checkForUTF8`, but this is relatively minor for this non-public function.

vadz commented on src/common/zipstrm.cpp:
```diff
@@ -2111,16 +2121,30 @@ size_t wxZipInputStream::OnSysRead(void *buffer, size_t size)
+                                     wxMBConv& conv)
   : wxArchiveOutputStream(stream, conv)
 {
     Init(level);
```
Shouldn't `m_useUTF8` be reset to false here, at least if `!conv.IsUTF8()`?

vadz reviewed (commented):
Thanks a lot!

I fully agree with the idea of this patch, we really should use UTF-8 by default. I do wonder why can't we just use `wxConvUTF8` as the default value for `conv` instead of `wxConvLocal`, wouldn't this be simpler than duplicating ctors etc?

We could make `IsUTF8()` always available by just removing `wxUSE_UNICODE_UTF8` guards around it.

paulcor commented:
Note that this is probably not a completely safe change. I don't know about other common zip tools, but the Info-ZIP tools do not recognize the 0x0800 UTF-8 flag. Won't make any difference with a UTF-8 locale, but on Windows it could cause problems for somebody. Not saying it shouldn't be done, just that there may be some negative consequences to consider.

vadz commented:
Thanks, I had no idea about this and naively thought that everybody supported UTF-8 nowadays (and, in fact, [Zip 3.0 release](http://www.info-zip.org/Zip.html) from 2008 does seem to include "UTF-8 filenames support", so does this only apply to older releases?). I still think it should be done though, but perhaps it should be noted that non-ASCII filenames risk creating some interoperability problems.

paulcor commented:
Info-ZIP does support UTF-8 paths. But it uses a different method, storing a second, UTF-8-encoded, path in an "extra field" record.

TcT2k committed:
Always enable wxMBConv::IsUTF8()

These where previously guarded by wxUSE_UNICODE_UTF8 but
may be useful in other configurations too.

TcT2k commented:
I've integrated @vadz suggestions and enabled `IsUTF8()` in all configurations which reduces changes required to implement UTF-8 ZIP creation.

I've included text in the documentation regarding incompatibilities which @paulcor mentioned.

vadz commented:
Thanks! Merged now, it would indeed be nice to have unit tests for this if you have time to write them.
